### PR TITLE
Added password support from nemccarthy

### PR DESCRIPTION
**Pull request by magnussand**
This is mearly a manually patched version from nemccarthy's pull request in order to get password support.

### CODE REVIEW

File: switchbot/__init__.py [hunk 1]
```diff
@@ -51,6 +62,19 @@ def _disconnect(self) -> None:
         finally:
             self._device = None
 
+    def _commandkey(self, key) -> str:
+        key = ""
```
Comment (Danielhiversen) [inline switchbot/__init__.py]: ??

File: switchbot/__init__.py [hunk 2]
```diff
@@ -51,6 +62,19 @@ def _disconnect(self) -> None:
...
+        key_suffix = PRESS_KEY_SUFFIX
+        if key == ON_KEY:
+            key_suffix = ON_KEY_SUFFIX
+        elif key == OFF_KEY:
+            key_suffix = OFF_KEY_SUFFIX
+        if self._password_encoded is not None:
+            key = KEY_PASSWORD_PREFIX + self._password_encoded + key_suffix
+        else:
+            key = KEY_PREFIX + key_suffix
+        return key
```
Comment (Danielhiversen) [inline switchbot/__init__.py]: Why is the key modified if no password is set?

**Comment by Danielhiversen** [comment]
@nemccarthy

File: switchbot/__init__.py (hunk 1 above)
Comment (magnussand) [reply switchbot/__init__.py]: Yes that is strange.

I'll look into it and get back

File: switchbot/__init__.py (hunk 2 above)
Comment (magnussand) [reply switchbot/__init__.py]: I'll get back to you when I've tested more

Comment (magnussand) [reply switchbot/__init__.py]: Good catch.

I've removed it and committed. Have tested, but note that the switchbot need to be set in switch mode or press mode from the app in order to use the press, turn_on and turn_off methods.

File: switchbot/__init__.py [hunk 3]
```diff
@@ -51,6 +62,18 @@ def _disconnect(self) -> None:
         finally:
             self._device = None
 
+    def _commandkey(self, key) -> str:
+        key_suffix = PRESS_KEY_SUFFIX
```
Comment (Danielhiversen) [inline switchbot/__init__.py]: I think this is better
```

       if self._password_encoded is None:
             return key

```

Comment (magnussand) [reply switchbot/__init__.py]: Pushed new version with proposed change

File: switchbot/__init__.py [hunk 4]
```diff
@@ -12,20 +12,31 @@
 UUID = "cba20d00-224d-11e6-9fb8-0002a5d5c51b"
 HANDLE = "cba20002-224d-11e6-9fb8-0002a5d5c51b"
 
+KEY_PREFIX = "5701"
```
Comment (Danielhiversen) [inline switchbot/__init__.py]: Unused

File: switchbot/__init__.py [hunk 5]
```diff
@@ -66,9 +87,11 @@ def _writekey(self, key) -> bool:
 
     def _sendcommand(self, key, retry) -> bool:
         send_success = False
+        command = self._commandkey(key)
+        _LOGGER.info("Sending command to switchbot %s", command)
```
Comment (Danielhiversen) [inline switchbot/__init__.py]: Change to debug

**Comment by Danielhiversen** [comment]
Thanks

**Comment by magnussand** [comment]
You're welcome!

I need this feature in Home Assistant since I don't want my neighbors to
easily be able to switch off my heating.

Are you the maintainer of the integration with HA?

Regards
Magnus

**Pull request merged**